### PR TITLE
profiler: Configurable profiling frequency

### PR DESCRIPTION
**Pull request by marselester**
Frederic proposed `--cpu-sampling-frequency` flag https://github.com/parca-dev/parca-agent/issues/1212.

### CODE REVIEW

File: cmd/parca-agent/main.go [hunk 1]
```diff
@@ -170,6 +182,11 @@ func main() {
 		}
 	}
 
+	if flags.CPUSamplingFrequency <= 0 {
+		level.Warn(logger).Log("msg", "cpu sampling frequency is too low. Setting it to the default value", "default", defaultCPUSamplingFrequency)
+		flags.CPUSamplingFrequency = defaultCPUSamplingFrequency
+	}
```
Comment (javierhonduco) [inline cmd/parca-agent/main.go:188]: Perhaps we could also have a warning if the frequency is bigger than some threshold, e.g. >150Hz?

**Review by javierhonduco — COMMENTED** [review]
Looking good!

File: cmd/parca-agent/main.go (hunk 1 above)
Comment (marselester) [reply cmd/parca-agent/main.go:188]: Sure 👍

File: README.md [hunk 2]
```diff
@@ -69,6 +69,9 @@ Flags:
                                   maps. 0 means no limit.
       --profiling-duration=10s    The agent profiling duration to use. Leave
                                   this empty to use the defaults.
+      --cpu-sampling-frequency=19
```
Comment (kakkoyun) [inline README.md:72]:
```suggestion
      --profiling-cpu-sampling-frequency=19
```
It'd be good to have a namespace/group. We have too many flags.

related https://github.com/parca-dev/parca-agent/pull/1075

File: pkg/profiler/cpu/cpu.go [hunk 3]
```diff
@@ -241,14 +244,18 @@ func (p *CPU) Run(ctx context.Context) error {
 		return fmt.Errorf("load bpf object: %w", err)
 	}
 
+	// Period is the number of events between sampled occurrences.
```
Comment (kakkoyun) [inline pkg/profiler/cpu/cpu.go:247]: ❤️

File: pkg/profiler/pprof.go [hunk 4]
```diff
@@ -21,7 +21,7 @@ import (
 )
 
 // ConvertToPprof converts several per process Profile to a pprof Profile.
-func ConvertToPprof(captureTime time.Time, prs ...*Profile) (*profile.Profile, error) {
+func ConvertToPprof(captureTime time.Time, periodns int64, prs ...*Profile) (*profile.Profile, error) {
```
Comment (kakkoyun) [inline pkg/profiler/pprof.go:24]:
```suggestion
func ConvertToPprof(captureTime time.Time, periodNS int64, prs ...*Profile) (*profile.Profile, error) {
```

File: docs/design.md [hunk 5]
```diff
@@ -1,6 +1,6 @@
 # Design
 
-Parca Agent implements a sampling profiler, to sample stack traces 100 times per second via eBPF. It tracks user space as well as kernel-space stack traces. From the raw data, it builds a [pprof](https://github.com/google/pprof) formatted profile and optionally sends it to a Parca server where it is stored and can be queried and analyzed over time.
+Parca Agent implements a sampling profiler, to sample stack traces 19 times per second via eBPF. It tracks user space as well as kernel-space stack traces. From the raw data, it builds a [pprof](https://github.com/google/pprof) formatted profile and optionally sends it to a Parca server where it is stored and can be queried and analyzed over time.
```
Comment (kakkoyun) [inline docs/design.md:3]: We should open an issue at https://github.com/parca-dev/docs to sync these before the next release.

https://github.com/parca-dev/docs/blob/cfcdda347b14d776467502dc3e2695983946b2d6/Makefile#L2

File: cmd/parca-agent/main.go [hunk 6]
```diff
@@ -76,6 +77,17 @@ const (
...
+	// We sample at 19Hz (19 times per second) because it is a prime number,
+	// and primes are good to avoid collisions with other things
+	// that may be happening periodically on a machine.
+	// In particular, 100 samples per second means every 10ms
+	// which is a frequency that may very well be used by user code,
+	// so a CPU profile could show a periodic workload on-CPU 100% of the time
+	// which is misleading.
+	defaultCPUSamplingFrequency = 19
+	// Setting the CPU sampling frequency too high can impact overall machine performance.
+	maxCPUSamplingFrequency = 150
```
Comment (kakkoyun) [inline cmd/parca-agent/main.go:90]:
```suggestion
	maxAdvicedCPUSamplingFrequency = 150
```
Let's be super pedantic about the naming here.

**Review by kakkoyun — COMMENTED** [review]
Great job! 

I have a few comments. Happy to merge it afterward. I'll give @brancz to have a look at it, though.

File: cmd/parca-agent/main.go [hunk 7]
```diff
@@ -170,6 +184,13 @@ func main() {
 		}
 	}
 
+	if flags.CPUSamplingFrequency <= 0 {
+		level.Warn(logger).Log("msg", "cpu sampling frequency is too low. Setting it to the default value", "default", defaultCPUSamplingFrequency)
+		flags.CPUSamplingFrequency = defaultCPUSamplingFrequency
+	} else if flags.CPUSamplingFrequency > maxCPUSamplingFrequency {
```
Comment (brancz) [inline cmd/parca-agent/main.go:190]: I think we discussed we also wanted to log if the frequency is not prime.

File: README.md [hunk 8]
```diff
@@ -4,7 +4,7 @@
 
 # Parca Agent
 
-Parca Agent is an always-on sampling profiler that uses eBPF to capture raw profiling data with very low overhead. It observes user-space and kernel-space stacktraces 100 times per second and builds [pprof](https://github.com/google/pprof) formatted profiles from the extracted data. Read more details in the [design documentation](docs/design.md).
+Parca Agent is an always-on sampling profiler that uses eBPF to capture raw profiling data with very low overhead. It observes user-space and kernel-space stacktraces 19 times per second and builds [pprof](https://github.com/google/pprof) formatted profiles from the extracted data. Read more details in the [design documentation](docs/design.md).
```
[the remark this answers is not in the document]
Comment (marselester) [reply README.md]: Thank you! Could you please update the diagram which is used in the design document https://docs.google.com/drawings/d/1Xq3VpXzO9wo2k91ZQKVBzzo4axszTA0SCrzRSnosNi4 when the PR is merged? It should state 19x per second.

Screenshot 2023-01-11 at 11 41 59

File: docs/design.md (hunk 5 above)
Comment (marselester) [reply docs/design.md]: Here we go https://github.com/parca-dev/docs/issues/212.

**Review by kakkoyun — APPROVED** [review]

**Pull request merged**